player/javascript.c

```diff
@@ -0,0 +1,1237 @@
+//
+// - push_success, push_failure, push_status and pushed_error set last error.
+
+// iserr as true indicates an error, and if so, str may indicate a reason.
+// Internally ctx->last_error_str is never NULL, and empty indicates success.
+static void set_last_error(struct script_ctx *ctx, bool iserr, const char *str)
+{
+    ctx->last_error_str[0] = 0;
+    if (!iserr)
+        return;
+    if (!str || !str[0])
+        str = "Error";
+    ctx->last_error_str = talloc_strdup_append(ctx->last_error_str, str);
+}
+
+// For use only by wrappers at defaults.js.
+// arg: error string. Use empty string to indicate success.
+static void script__set_last_error(js_State *J)
+{
+    const char *e = js_tostring(J, 1);
+    set_last_error(jctx(J), e[0], e);
+}
+
+// mp.last_error() . args: none. return the last error without modifying it.
+static void script_last_error(js_State *J)
+{
+    js_pushstring(J, jctx(J)->last_error_str);
+}
+
+// Generic success for APIs which don't return an actual value.
+static void push_success(js_State *J)
+{
+    set_last_error(jctx(J), 0, NULL);
+    js_pushboolean(J, true);
+}
+
+// Doesn't (intentionally) throw. Just sets last_error and pushes undefined
+static void push_failure(js_State *J, const char *str)
+{
+    set_last_error(jctx(J), 1, str);
+    js_pushundefined(J);
+}
+
+// Most of the scripting APIs are either sending some values and getting status
+// code in return, or requesting some value while providing a default in case an
+// error happened. These simplify the C code for that and always set last_error.
+
+static void push_status(js_State *J, int err)
+{
+    if (err >= 0) {
+        push_success(J);
+    } else {
+        push_failure(J, mpv_error_string(err));
+    }
+}
+
+ // If err is success then return 0, else push the item at def and return 1
+static bool pushed_error(js_State *J, int err, int def)
+{
+    bool iserr = err < 0;
+    set_last_error(jctx(J), iserr, iserr ? mpv_error_string(err) : NULL);
+    if (!iserr)
+        return false;
+
+    js_copy(J, def);
+    return true;
+}
+
+/**********************************************************************
+ *  Autowash - care-free resource deallocation on vm errors, and otherwise
+ *********************************************************************/
+// - Autowash (aw) functions are called with a pre-allocated and zeroed autowash
+//   object argument which is also a talloc context. The object's members and
+//   the object itself are freed after the function exits - either normally or
+//   because it threw an error, on the latter case it then re-throws the error.
+//
+//   Functions with autowash* argument should be inserted as js C functions with
+//   aw_newcfunction and used normally, or called from C with the AW macro.
+
+typedef struct autowash {
+    char *ma;  // mpv alloc -> mpv_free(..)
+    mpv_node result_node;  // a node obtained via an mpv API
+    FILE *f;   // fclose
+    DIR *dir;  // closedir
```

Comment:
Using opaque pointers with destructors is only slightly more elegant IMO, but still means it's limited to the number of pointers the struct has, if I understood you correctly.

If we really want a generic approach, then IMO the `af` (previously `aw`) pointer should just be an empty talloc context, where we add to it as many destructed items was we like. For instance this patch does it for the `FILE*` member:

```diff
diff --git a/player/javascript.c b/player/javascript.c
index 91a948d..92e3c3e 100644
--- a/player/javascript.c
+++ b/player/javascript.c
@@ -194,10 +194,24 @@ static bool pushed_error(js_State *J, int err, int def)
 //   Functions with autofree* argument should be inserted as js C functions with
 //   af_newcfunction and used normally, or called from C with the AF macro.
 
+
+static void delete_af_file(void *p)
+{
+    FILE *f = *(FILE**)p;
+    if (f)
+        fclose(f);
+}
+
+static FILE **new_af_file(void *parent)
+{
+    FILE **f = talloc_zero(parent, FILE*);
+    talloc_set_destructor(f, delete_af_file);
+    return f;
+}
+
 typedef struct autofree {
     char *ma;  // mpv alloc -> mpv_free(..)
     mpv_node result_node;  // a node obtained via an mpv API
-    FILE *f;   // fclose
     DIR *dir;  // closedir
 } autofree;
 
@@ -213,8 +227,6 @@ static void delete_autofree(autofree *af) {
     // "If node->format is MPV_FORMAT_NONE, this call does nothing." (client.h)
     if (af->result_node.format != MPV_FORMAT_NONE)
         mpv_free_node_contents(&af->result_node);
-    if (af->f)
-        fclose(af->f);
     if (af->dir)
         closedir(af->dir);
 
@@ -298,7 +310,7 @@ static void af_push_file(js_State *J, const char *fname, int limit, autofree *af
         return;
     }
 
-    FILE *f = af->f = fopen(filename, "rb");
+    FILE *f = *new_af_file(af) = fopen(filename, "rb");
     if (!f)
         js_error(J, "cannot open file: '%s'", filename);
 
@@ -913,7 +925,7 @@ static void script_write_file(js_State *J, autofree *af)
     fname = mp_get_user_path(af, jctx(J)->mpctx->global, fname);
     MP_VERBOSE(jctx(J), "Writing file '%s'\n", fname);
 
-    FILE *f = af->f = fopen(fname, "wb");
+    FILE *f = *new_af_file(af) = fopen(fname, "wb");
     if (!f)
         js_error(J, "Cannot open file for writing: '%s'", fname);
```

However, I don't think it makes the code more readable or otherwise much better. I do agree that if the struct started growing more members then a generic approach would be better at some stage, but for now, all it needs is 4 members.

Anyway, if you prefer this approach, I'll apply it to the other members too. Let me know.